onnx/src/ops/ml/mod.rs

```diff
@@ -0,0 +1,330 @@
+                Node::new_leaf(start, end)
+            } else {
+                node.bail(&format!("leaf not found: tree_id={}, node_id={}", tree_id, node_id))?
+            }
+        };
+        nodes[tree_id].push(tree_node);
+
+        prev_tree_id = tree_id;
+        prev_node_id = node_id;
+    }
+
+    // build the trees
+
+    let mut trees: Vec<Tree> = Vec::default();
+
+    for i in 0..n_trees {
+        let tree = Tree::build(n_classes, &nodes[i], &leaves[i])
+            .or_else(|err| node.bail(&format!("{}", err)))?;
+        trees.push(tree);
+    }
+
+    // build the tree ensemble
+
+    let base_scores = base_values.as_ref().map(Vec::as_slice);
+    let ensemble = TreeEnsemble::build(&trees, aggregate_fn, post_transform, base_scores)
+        .or_else(|err| node.bail(&format!("{}", err)))?;
+
+    Ok(ensemble)
+}
+
+pub fn register_all_ops(reg: &mut OpRegister) {
+    reg.insert("TreeEnsembleClassifier", TreeEnsembleClassifier::parse);
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub enum ClassLabels {
+    Ints(Vec<i64>),
+    Strings(Vec<String>),
+}
+
+#[derive(Debug, Clone)]
+pub struct TreeEnsembleClassifier {
+    ensemble: TreeEnsemble,
+    class_labels: ClassLabels,
+}
+
+impl TreeEnsembleClassifier {
+    fn parse(node: &NodeProto) -> TractResult<Box<Op>> {
+        let ensemble = parse_tree_ensemble(node, true)?;
+        let class_labels = match node.get_attr_opt_slice::<i64>("classlabels_int64s")? {
+            Some(int_labels) => ClassLabels::Ints(int_labels.into()),
+            _ => {
+                let str_labels = node.get_attr_opt_vec::<String>("classlabels_strings")?;
+                ClassLabels::Strings(str_labels.unwrap())
+            }
+        };
+        Ok(Box::new(Self { ensemble, class_labels }))
+    }
+}
+
+impl Op for TreeEnsembleClassifier {
+    fn name(&self) -> Cow<str> {
+        "TreeEnsembleClassifier".into()
+    }
+
+    fn noutputs(&self) -> usize {
+        2
+    }
+}
+
+impl StatelessOp for TreeEnsembleClassifier {
+    fn eval(&self, mut inputs: TVec<SharedTensor>) -> TractResult<TVec<SharedTensor>> {
+        Err("WIP".into())
+    }
+}
+
+impl InferenceRulesOp for TreeEnsembleClassifier {
+    fn rules<'r, 'p: 'r, 's: 'r>(
+        &'s self, s: &mut Solver<'r>, inputs: &'p [TensorProxy], outputs: &'p [TensorProxy],
+    ) -> InferenceResult {
+        check_input_arity(&inputs, 1)?;
+        check_output_arity(&outputs, 2)?;
+
+        s.equals(&inputs[0].rank, 2)?;
+        s.equals(&outputs[0].rank, 1)?;
+        s.equals(&outputs[0].rank, 1)?;
+
+        s.given_3(
+            &inputs[0].shape,
+            &outputs[0].shape,
+            &outputs[1].shape,
+            move |s, s_in0, s_out0, s_out1| {
+                s.equals(s_in0[0], s_out0[0])?;
+                s.equals(s_out0[0], s_out1[0])?;
+                if let Ok(k) = s_in0[1].to_integer() {
+                    self.ensemble.check_n_features(k as _)?;
+                }
```

Comment:
These two checks are range checks, so going for "given" is the right option. In an ideal world, we would have rules in the form of "a gt b" and "a lte b", but implementing reasoning on range feels overkill.

Reply:
Ha, I was almost about to ask, why not have `s.lt()` and `s.le()` (maybe also `s.not()` then) :)